File: decompiled/c/419f50.c

void FUN_00419f50(undefined2 *param_1,uint param_2,int param_3)

{
  uint uVar1;
  undefined4 *puVar2;
  int iVar3;
  undefined2 local_c;
  undefined4 uStack_a;
  undefined2 uStack_6;
  undefined4 local_4;
  
  iVar3 = 0x5eeb18;
  if (param_2 != 0) {
    if ((int)param_2 < 0) {
      param_2 = -param_2;
      iVar3 = 0x5eec78;
    }
    if (param_3 == 0) {
      *param_1 = 0;
    }
    while (param_2 != 0) {
      iVar3 = iVar3 + 0x54;
      uVar1 = param_2 & 7;
      param_2 = (int)param_2 >> 3;
      if (uVar1 != 0) {
        puVar2 = (undefined4 *)(iVar3 + uVar1 * 0xc);
        if (0x7fff < *(ushort *)(iVar3 + uVar1 * 0xc)) {
          local_c = (undefined2)*puVar2;
          uStack_a._0_2_ = (undefined2)((uint)*puVar2 >> 0x10);
          uStack_a._2_2_ = (undefined2)puVar2[1];
          uStack_6 = (undefined2)((uint)puVar2[1] >> 0x10);
          local_4 = puVar2[2];
          uStack_a = CONCAT22(uStack_a._2_2_,(undefined2)uStack_a) + -1;
          puVar2 = (undefined4 *)&local_c;
        }
        FUN_00419c90(param_1,puVar2);
      }
    }
  }
  return;
}

File: ported/auto/419f50.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/419f50.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { CONCAT22 } from "../../runtime/ghidra-builtins.js";
import { regs } from "../../runtime/regs.js";
import { FUN_00419c90 } from "./419c90.js";
export function FUN_00419f50(heap, param_1, param_2, param_3) {
  const __sp = heap.allocFrame(12);
  const __addr_local_c = __sp + 0;
  const __addr_local_4 = __sp + 8;
  try {
  let uVar1 = 0;
  let puVar2 = 0;
  let iVar3 = 0;
  let uStack_a = 0;
  let uStack_6 = 0;
  iVar3 = ((0x5eeb18) >>> 0);
  if (param_2 != 0) {
    if (((param_2) | 0) < 0) {
      param_2 = ((-param_2) >>> 0);
      iVar3 = ((0x5eec78) >>> 0);
    }
    if (param_3 == 0) {
      heap.setU32(param_1, (0) & 0xffffffff);
    }
    while (param_2 != 0) {
      iVar3 = ((iVar3 + 0x54) >>> 0);
      uVar1 = ((param_2 & 7) >>> 0);
      param_2 = ((((param_2) | 0) >>> 3) >>> 0);
      if (uVar1 != 0) {
        puVar2 = (((iVar3 + uVar1 * 0xc)) >>> 0);
        if (0x7fff < heap.u16((iVar3 + uVar1 * 0xc))) {
          heap.setU32(__addr_local_c, (heap.u16(puVar2)) >>> 0);
          (function(){ throw new Error("c-to-js: unhandled LHS form field_expression in FUN_00419f50"); })();
          (function(){ throw new Error("c-to-js: unhandled LHS form field_expression in FUN_00419f50"); })();
          uStack_6 = ((((((heap.u32(puVar2 + (1) * 4)) >>> 0) >>> 0x10) & 0xffff)) & 0xffff);
          heap.setU32(__addr_local_4, (heap.u32(puVar2 + (2) * 4)) >>> 0);
          uStack_a = ((CONCAT22((((uStack_a) >>> 16) & 0xffff), ((uStack_a) & 0xffff)) + -1) >>> 0);
          puVar2 = ((__addr_local_c) >>> 0);
        }
        (regs.eax = FUN_00419c90(heap, param_1, puVar2));
      }
    }
  }
  return;
} finally {
    heap.freeFrame(12);
  }
}
